js/modules/k6/websocket/websocket.go

```diff
@@ -0,0 +1,228 @@
+/*
+ *
+ * k6 - a next-generation load testing tool
+ * Copyright (C) 2017 Load Impact
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package websocket
+
+import (
+	"bytes"
+	"context"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+
+	"github.com/dop251/goja"
+	"github.com/gorilla/websocket"
+	"github.com/loadimpact/k6/js/common"
+)
+
+type Websocket struct{}
+
+type Socket struct {
+	ctx           context.Context
+	conn          *websocket.Conn
+	eventHandlers map[string]goja.Callable
+	scheduled     chan goja.Callable
+	done          chan struct{}
+}
+
+const writeWait = 10 * time.Second
+
+var (
+	newline = []byte{'\n'}
+	space   = []byte{' '}
+)
+
+func (*Websocket) Connect(ctx context.Context, url string, setupFn goja.Callable) {
+	rt := common.GetRuntime(ctx)
+	conn, _, connErr := websocket.DefaultDialer.Dial(url, nil)
+
+	socket := Socket{
+		ctx:           ctx,
+		conn:          conn,
+		eventHandlers: make(map[string]goja.Callable),
+		scheduled:     make(chan goja.Callable),
+		done:          make(chan struct{}),
+	}
+
+	// Run the user-provided set up function
+	setupFn(goja.Undefined(), rt.ToValue(&socket))
+
+	if connErr != nil {
+		// Pass the error to the user script before exiting immediately
+		socket.handleEvent("error", rt.ToValue(connErr))
+		return
+	}
+	defer conn.Close()
+
+	conn.SetPongHandler(func(string) error { socket.handleEvent("pong"); return nil })
+
+	// The connection is now open, emit the event
+	socket.handleEvent("open")
+
+	readDataChan := make(chan []byte)
+	readErrChan := make(chan error)
+	interrupt := make(chan os.Signal, 1)
+	signal.Notify(interrupt, os.Interrupt, syscall.SIGINT, syscall.SIGTERM)
```

Comment:
Instead of a separate interrupt channel, use `ctx.Done()`; listening for signals will keep the socket running even if the test is scaled down and the VU suspended, or the test is stopped through the API.